### PR TITLE
Make LU gradient work for low-rank matrices

### PR DESCRIPTION
**Pull request by dpfau**
This PR makes it so the gradient of the LU decomposition can be correctly computed for low-rank matrices. This introduces an additional computational overhead of one extra triangular_solve - each triangular_solve scales as O(n^2), while LU itself scales as O(n^3), so the additional overhead should be minimal.

This makes it so the second derivatives of the determinant of corank-1 matrices can be computed correctly. There are further issues with corank >1 matrices that this PR does not address.

One possible issue is that the behavior of the L term in the LU decomposition is not defined for low-rank matrices for columns greater than the rank of the matrix, and in fact behaves differently for different backends (certain terms that are 0 on CPU appear as 0.5 on GPU on the corank-2 matrix used in the tests, for instance). At the moment the behavior of the gradient of the L matrix for columns beyond the rank of the input matrix is not checked in the tests. As the LU decomposition is mainly used as an inner loop for matrix solves and determinants, we hope this will not introduce any undue confusion in the future.

### CODE REVIEW

File: jax/lax_linalg.py [hunk 1]
```diff
@@ -612,13 +612,30 @@ def _lu_jvp_rule(primals, tangents):
   u_padding[-2] = (0, n - k, 0)
   u = lax.pad(jnp.triu(lu[..., :k, :]), zero, u_padding) + u_eye
 
-  la = triangular_solve(l, x, left_side=True, transpose_a=False, lower=True,
-                        unit_diagonal=True)
-  lau = triangular_solve(u, la, left_side=False, transpose_a=False,
-                         lower=False)
+  # dummy calculation to figure out which matrices u are low-rank
```
Comment (shoyer) [inline jax/lax_linalg.py:615]: Are you sure we need a triangular solve for this? Given that u is triangular, can’t we just check if one of the diagonal entries is zero?

File: tests/linalg_test.py [hunk 2]
```diff
@@ -42,6 +42,29 @@
 float_types = [np.float32, np.float64]
 complex_types = [np.complex64, np.complex128]
 
+
+# Used in tests for det and lu
+singular_mats = [
```
Comment (shoyer) [inline tests/linalg_test.py:47]: Nit: I think it’s a best practice to keep data for data dependent tests defined inside or directly adjacent to the test method?

File: tests/linalg_test.py [hunk 3]
```diff
@@ -973,6 +993,33 @@ def testLuGrad(self, shape, dtype, rng_factory):
...
+      for corank in [1, 2]))
+  @jtu.skip_on_devices("tpu")  # TODO(mattjj, pfau): fails on TPU.
+  def testLuGradOfSingularMatrix(self, corank):
+    def _lu(a):
+      p, l, u = jsp.linalg.lu(a)
+      return p, l[:, :-corank], u
+    mat = singular_mats[corank-1]
+    # Check forward mode gradients
+    jtu.check_grads(_lu, (mat,), 2, modes=['fwd'], atol=1e-1, rtol=1e-1)
+    # Something is causing tests to fail for reverse-mode.
```
Comment (shoyer) [inline tests/linalg_test.py:1007]: I suspect that the automated reverse mode check fails because it passes in a cotangent with nonzero elements outside of the triangular matrix.

Can you instead make the `_lu` helper return only the lower/upper triangular parts of the results? I.e., the part that is allowed to take on nonzero values?

File: jax/lax_linalg.py [hunk 4]
```diff
@@ -612,13 +612,30 @@ def _lu_jvp_rule(primals, tangents):
...
+  u_safe = jnp.where(jnp.swapaxes(u_mask, -1, -2), 1.0, u)
+  la = triangular_solve(
+      l, x, left_side=True, transpose_a=False, lower=True, unit_diagonal=True)
+  lau = triangular_solve(
+      u_safe, la, left_side=False, transpose_a=False, lower=False)
+  lau = jnp.where(mask, 0.0, lau)
 
   l_dot = jnp.matmul(l, jnp.tril(lau, -1))
   u_dot = jnp.matmul(jnp.triu(lau), u)
+  # Correction for low-rank matrices
```
Comment (shoyer) [inline jax/lax_linalg.py:634]: This correction is not intuitive to me. Could you add a comment noting what’s going on here?

File: jax/lax_linalg.py [hunk 5]
```diff
@@ -612,13 +612,30 @@ def _lu_jvp_rule(primals, tangents):
...
+  u_fix = la - triangular_solve(
+      l, l_dot @ u,
```
Comment (shoyer) [inline jax/lax_linalg.py:636]: This is a bit of a gotcha with JAX currently, but on TPUs the default matmul precision is currently bfloat16, which is generally entirely insufficient for linear algebra. Can you use jnp.dot here with precision=lax.Precision.HIGHEST? TPU performance for matrix decompositions is currently pretty awful, but at least we can ensure that it gives correct results...

File: jax/lax_linalg.py (hunk 1 above)
Comment (dpfau) [reply jax/lax_linalg.py:615]: This was something we discussed when updating the gradient of `jnp.linalg.det` to work for low-rank matrices as well. There are issues with underflow so that you sometimes get infs/nans even if the diagonal isn't exactly zero. Consider this example:

```
import jax
impor jax.numpy as jnp
import struct

iota = struct.unpack('!f', bytes.fromhex('00000001'))[0]
mat = jnp.array([[iota, 0], [0, iota]])
```

If you do `iota=0` it will return False, but if you do `jax.lax_linalg.triangular_solve(mat, jnp.ones((2, 1)), left_side=True)` it will return

```
DeviceArray([[nan],
             [inf]], dtype=float32)
```

File: tests/linalg_test.py (hunk 2 above)
Comment (dpfau) [reply tests/linalg_test.py:47]: I can move the non-square matrices, but `singular_mats` is used by two tests (for both determinants and LU decomposition). What's the best practice there?

File: jax/lax_linalg.py (hunk 5 above)
Comment (dpfau) [reply jax/lax_linalg.py:636]: Should I do that throughout the function, or just replace `l_dot @ u` with `jnp.dot(l_dot, u, precision=lax.Precision.HIGHEST)`? Does it matter if it's `dot` or `matmul`?

Comment (dpfau) [reply jax/lax_linalg.py]: I've changed this line. Let me know if you think I should apply the same change elsewhere.

File: tests/linalg_test.py (hunk 3 above)
Comment (dpfau) [reply tests/linalg_test.py:1007]: I don't think this is the issue, since `jsp.linalg.lu` masks out the upper and lower triangulars with zero anyway, and `jacfwd` and `jacrev` match.

File: tests/linalg_test.py (hunk 2 above)
Comment (dpfau) [reply tests/linalg_test.py:47]: I moved nonsquare_singular_mats.

Comment (dpfau) [reply tests/linalg_test.py:47]: The linter keeps complaining now and the tests are failing. I've reverted back to the way it was - at least that worked.

File: tests/linalg_test.py (hunk 3 above)
Comment (dpfau) [reply tests/linalg_test.py:1007]: The test that's here now - looking at the entire Jacobian in forward and reverse mode, instead of just a random projection - is actually a stronger test that `jtu.check_grads`. So I'm inclined to leave it.

File: jax/lax_linalg.py (hunk 1 above)
Comment (shoyer) [reply jax/lax_linalg.py:615]: What about using `jnp.diag(u) < jnp.ffinfo(u.dtype).eps` (or some multiple of `eps`)? I guess the `triangular_solve` is probably fine (and measures exactly what we care about).

File: tests/linalg_test.py (hunk 3 above)
Comment (shoyer) [reply tests/linalg_test.py:1007]: Hmm, I just tried this out and you may be right.

I'm a little nervous that we aren't testing higher order derivatives in reverse mode, but this is definitely an improvement.

File: jax/lax_linalg.py (hunk 4 above)
Comment (shoyer) [reply jax/lax_linalg.py:634]: thanks! I mostly follow this now :)

**Comment by shoyer** [comment]
One last thing before we merge this -- I agree with your theoretical analysis that the slow down should be negligible, but it would be nice to some sort of empirical microbenchmark as well.

File: jax/lax_linalg.py (hunk 1 above)
Comment (dpfau) [reply jax/lax_linalg.py:615]: I'm just worried that there may be subtle edge cases where 1/eps is finite but whatever it is we're multiplying by isn't, or different backends may have subtly different behavior, and basically the safest way is to just invert it directly. I should also say that because we are only doing a triangular solve against a vector, not a matrix, it will run faster than the other triangular solves in this function.

File: jax/lax_linalg.py (hunk 4 above)
Comment (dpfau) [reply jax/lax_linalg.py:634]: Yeah, the full derivation is a little too long for the margin of this comment ;)

File: jax/lax_linalg.py (hunk 1 above)
Comment (shoyer) [reply jax/lax_linalg.py:615]: Great point on the vector vs. matrix valued solve!

**Comment by dpfau** [comment]
Results from some internal tests:

```
CPU:

Shape:  (10000, 10, 10)
Time for old JVP rule (sec):  0.0672062603989616
Time for new JVP rule (sec):  0.09585398440249264

Shape:  (1000, 100, 100)
Time for old JVP rule (sec):  1.786021307995543
Time for new JVP rule (sec):  3.2155005412874744

Shape:  (10, 1000, 1000)
Time for old JVP rule (sec):  0.9578036440070719
Time for new JVP rule (sec):  1.247297911881469

Shape:  (2000, 2000)
Time for old JVP rule (sec):  0.9859540804987773
Time for new JVP rule (sec):  1.3910707260016353

GPU:

Shape:  (10000, 10, 10)
Mean time for old JVP rule (sec):  0.0020070552825927736
Mean time for new JVP rule (sec):  0.002073383331298828

Shape:  (1000, 100, 100)
Mean time for old JVP rule (sec):  0.08613080978393554
Mean time for new JVP rule (sec):  0.191933536529541

Shape:  (10, 1000, 1000)
Mean time for old JVP rule (sec):  0.17102200984954835
Mean time for new JVP rule (sec):  0.18128788471221924

Shape:  (2000, 2000)
Mean time for old JVP rule (sec):  0.10356688499450684
Mean time for new JVP rule (sec):  0.10452637672424317

TPU:

Shape:  (10000, 10, 10)
Time for old JVP rule (sec):  0.0015675264159217476
Time for new JVP rule (sec):  0.0018886566990986467

Shape:  (1000, 100, 100)
Time for old JVP rule (sec):  0.09939546976238489
Time for new JVP rule (sec):  0.18620093669742346

Shape:  (10, 1000, 1000)
Time for old JVP rule (sec):  0.1820111739449203
Time for new JVP rule (sec):  0.19434180669486523

Shape:  (2000, 2000)
Time for old JVP rule (sec):  0.12048129616305232
Time for new JVP rule (sec):  0.12088406663388014
```

**Review by shoyer — APPROVED** [review]

**Comment by shoyer** [comment]
OK, cool.  We should seriously consider adding an optional keyword argument to control the gradient implementation (https://github.com/google/jax/issues/3641), but for now I would definitely prioritize correctness over speed.

File: tests/linalg_test.py [hunk 6]
```diff
@@ -42,6 +42,29 @@
...
+
+nonsquare_singular_mats = [
+  jnp.array([[-35,   7,  27, -17],
+             [ 11, -13,  -9,  11],
+             [ 19, -11, -15,  13]], dtype=jnp.float32),
+  jnp.array([[-35,  11,  19],
+             [  7, -13, -11],
+             [ 27,  -9, -15],
+             [-17,  11,  13]], dtype=np.float32)
+]
```
Comment (jakevdp) [inline tests/linalg_test.py:65]: This PR broke internal tests because the device is not yet initialized when these lines are run. We need to replace these module-level constants with functions.

Comment (hawkinsp) [reply tests/linalg_test.py:65]: You can also often use classic NumPy arrays in this situation.

Comment (jakevdp) [reply tests/linalg_test.py:65]: Should be fixed by #3684

**Comment by shoyer** [comment]
Ah, rats! I should have known to run internal tests before merging.

**Comment by jakevdp** [comment]
I ended up rolling-back this change in #3688, because it caused nontrivial test failures internally on CPU, GPU, and TPU. I haven't had the chance to dig into why it failed there when it passed on travis, but we should be sure to run internal tests before re-landing this.

**Comment by dpfau** [comment]
Not sure I have time to dig into this right now, but I'll fix this as part of implementing the `grad_type` flag for LU and determinant gradients as in #3641

**Comment by jakevdp** [comment]
Thanks! Sorry for the rollback, and let me know if I can help out.

**Comment by dpfau** [comment]
From a quick look at the test results - I think what's happening is that rows of dL are getting mixed up, because the ordering is arbitrary for low-rank matrices. Not sure how to fix it, but at least I think that's what's happening. Also the leftmost columns are different, but as I mention in the comments, there's no way to avoid that because different backends actually treat the leftmost columns of L differently for low rank matrices.

```
 x: array([[ 0.000000e+00,  0.000000e+00,  0.000000e+00],
       [-7.410581e-02,  0.000000e+00,  0.000000e+00],
       [ 1.274934e-02, -1.163514e-01,  0.000000e+00],
       [-6.602654e-02, -6.963590e-02, -5.648539e+06]], dtype=float32)
 y: array([[ 0.      ,  0.      ,  0.      ],
       [-0.074059,  0.      ,  0.      ],
       [-0.066161, -0.069514,  0.      ],
       [ 0.012517, -0.116527,  6.867051]], dtype=float32)
```

**Comment by dpfau** [comment]
But I have absolutely no idea what is up with the test on TPU. Maybe best to disable that.
